lib/src/main/java/tech/relaycorp/relaydroid/messaging/IncomingMessage.kt

```diff
@@ -19,15 +20,38 @@ public class IncomingMessage internal constructor(
 ) {
 
     internal companion object {
-        internal suspend fun build(parcel: Parcel, ack: suspend () -> Unit) =
-            IncomingMessage(
+        @Throws(
+            UnknownFirstPartyEndpointException::class,
+            UnknownThirdPartyEndpointException::class,
+            PersistenceException::class
+        )
+        internal suspend fun build(parcel: Parcel, ack: suspend () -> Unit): IncomingMessage {
+            val recipientEndpoint = FirstPartyEndpoint.load(parcel.recipientAddress)
+                ?: throw UnknownFirstPartyEndpointException(
+                    "Unknown third party endpoint with address ${parcel.recipientAddress}"
+                )
+
+            val sender = ThirdPartyEndpoint.load(
+                parcel.recipientAddress,
+                parcel.senderCertificate.subjectPrivateAddress,
+            ) ?: throw UnknownThirdPartyEndpointException(
+                "Unknown third party endpoint with address " +
+                    "${parcel.senderCertificate.subjectPrivateAddress} " +
+                    "for first party endpoint ${parcel.recipientAddress}"
+            )
+
+            return IncomingMessage(
                 id = MessageId(parcel.id),
                 payload = parcel.payload,
-                senderEndpoint = PrivateThirdPartyEndpoint(parcel.senderCertificate.subjectPrivateAddress),
-                recipientEndpoint = FirstPartyEndpoint.load(parcel.recipientAddress)!!,
+                senderEndpoint = sender,
+                recipientEndpoint = recipientEndpoint,
                 creationDate = parcel.creationDate,
                 expiryDate = parcel.expiryDate,
                 ack = ack
             )
+        }
     }
 }
+
+public class UnknownThirdPartyEndpointException(message: String) : RelaynetException(message, null)
+public class UnknownFirstPartyEndpointException(message: String) : RelaynetException(message, null)
```

Comment:
Why don't we throw these exceptions in `[Third|First]PartyEndpoint.load()` already, instead of returning `null`?

Reply:
I thought about that, but I didn't want to return exceptions for something that might not be unexpected. We might use load just to check if something is there. Exception can ruin a bit of flow in the code, so I try to use them a bit sparingly. And we already have a lot of them :P

Reply:
Fair point